docs/architecture/adr-008-app-caller-cbs/adr-008-app-caller-cbs.md

```diff
@@ -0,0 +1,376 @@
+    // may return the empty string
+    GetSrcCallbackAddress() string
+
+    // may return the empty string
+    GetDestCallbackAddress() string
+
+    // UserDefinedGasLimit allows the sender of the packet to define inside the packet data
+    // a gas limit for how much the ADR-8 callbacks can consume. If defined, this will be passed
+    // in as the gas limit so that the callback is guaranteed to complete within a specific limit.
+    // On recvPacket, a gas-overflow will just fail the transaction allowing it to timeout on the sender side.
+    // On ackPacket and timeoutPacket, a gas-overflow will reject state changes made during callback but still
+    // commit the transaction. This ensures the packet lifecycle can always complete.
+    // If the packet data returns 0, the remaining gas limit will be passed in (modulo any chain-defined limit)
+    // Otherwise, we will set the gas limit passed into the callback to the `min(ctx.GasLimit, UserDefinedGasLimit())`
+    UserDefinedGasLimit() uint64
+}
+```
+
+IBC Apps or middleware can then call the IBCActor callbacks like so in their own callbacks:
+
+### Handshake Callbacks
+
+The handshake init callbacks (`OnChanOpenInit` and `OnChanCloseInit`) will need to include an additional field so that the initiating actor can be tracked and called upon during handshake completion.
+
+```go
+func OnChanOpenInit(
+    ctx sdk.Context,
+    order channeltypes.Order,
+    connectionHops []string,
+    portID string,
+    channelID string,
+    channelCap *capabilitytypes.Capability,
+    counterparty channeltypes.Counterparty,
+    version string,
+    actor string,
+) (string, error) {
+    acc := k.getAccount(ctx, actor)
+    ibcActor, ok := acc.(IBCActor)
+    if ok {
+        k.setActor(ctx, portID, channelID, actor)
+    }
+    
+    // continued logic
+}
+
+func OnChanOpenAck(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+    counterpartyChannelID string,
+    counterpartyVersion string,
+) error {
+    // run any necessary logic first
+    // negotiate final version
+
+    actor := k.getActor(ctx, portID, channelID)
+    if actor != "" {
+        ibcActor, _ := acc.(IBCActor)
+        ibcActor.OnChanOpen(ctx, portID, channelID, version)
+    }
+    // cleanup state
+    k.deleteActor(ctx, portID, channelID)
+}
+
+func OnChanOpenConfirm(
+    ctx sdk.Context,
+    portID,
+    channelID string,
+) error {
+    // run any necesssary logic first
+    // retrieve final version
+
+    actor := k.getActor(ctx, portID, channelID)
+    if actor != "" {
+        ibcActor, _ := acc.(IBCActor)
+        ibcActor.OnChanOpen(ctx, portID, channelID, version)
+    }
+    // cleanup state
+    k.deleteActor(ctx, portID, channelID)
+}
+
+func OnChanCloseInit(
+    ctx sdk.Context,
+    portID,
+    channelID,
+    actor string,
+) error {
+    acc := k.getAccount(ctx, actor)
+    ibcActor, ok := acc.(IBCActor)
+    if ok {
+        k.setActor(ctx, portID, channelID, actor)
+    }
```

Comment:
seems odd to set a different actor for the closing handshake steps rather than just reusing what's already there. i.e. not deleting the state entry in ACK/CONFIRM.